### PR TITLE
More predictive order for what IP address is chosen for device connections (use link local as last resort)

### PR DESCRIPTION
**Pull request by andy31415**
#### Problem
I have Unifi mDNS reflector enabled on my network. The side-effect is that all mdns queries and responses are seen throughout my devices, including my eth which cannot talk to link-local addreses on a esp32 wifi endpoint.

#### Change overview
Sort mDNS ip addresses as "IPv6 non-link-local first, then IPv4, then IPv6 LL". Currently we communicate just using the first address, so effectively this makes LL addresses as a last resort (if device has nothing else).

#### Testing
Use mdns reflector, `onoff toggle` failing more often than not, observed IP address used as a LL on a n interface that cannot use that LL. After this change, onoff toggle succeeded several times in a row (>5), never failed.

### CODE REVIEW

**Review by yunhanw-google — APPROVED** [review]

**Review by cecille — APPROVED** [review]

**Review by bzbarsky-apple — COMMENTED** [review]
This seems to be a 1-2KB codesize increase.  Is that expected?

File: src/lib/dnssd/Resolver.h [hunk 1]
```diff
@@ -17,6 +17,7 @@
 
 #pragma once
 
+#include <algorithm>
```
Comment (bzbarsky-apple) [inline src/lib/dnssd/Resolver.h:20]: I assume this is what causes the big increase in codesize....

**Comment by andy31415** [comment]
> This seems to be a 1-2KB codesize increase. Is that expected?

Not really,  this is likely sorting 2-3 IP addresses only. I  saw algorithm already used in other places, so I included it. I will  try it with a manually slow (but good enough on small data sets) implementation.

File: src/lib/dnssd/Resolver.h (hunk 1 above)
Comment (andy31415) [reply src/lib/dnssd/Resolver.h]: Updated to not use algorithm. Still used std::swap since that seemed very safe.

**Review by bzbarsky-apple — APPROVED** [review]

**Comment by andy31415** [comment]
After the new change, it seems better:

```
bloaty out/efr32-brd4161a-window-covering/chip-efr32-window-example.out -- out-master/efr32-brd4161a-window-covering/chip-efr32-window-example.out
    FILE SIZE        VM SIZE
 --------------  --------------
  +0.0% +1.01Ki  [ = ]       0    .debug_info
  +0.0%    +359  [ = ]       0    .debug_str
  +0.0%    +224  +0.0%    +224    .text
  +0.0%    +141  [ = ]       0    .debug_line
  +0.0%    +120  [ = ]       0    .debug_loc
  +0.0%     +66  [ = ]       0    .debug_abbrev
  +0.0%     +48  [ = ]       0    .symtab
  +0.0%     +34  [ = ]       0    .strtab
  +0.0%     +32  [ = ]       0    .debug_ranges
  +0.0%      +4  [ = ]       0    .debug_frame
  -0.9%      -2  [ = ]       0    .shstrtab
  -0.3%    -224  [ = ]       0    .nvm_dummy
  -0.3%    -224  [ = ]       0    [Unmapped]
  +0.0% +1.57Ki  +0.0%    +224    TOTAL
  ```

**Comment by andy31415** [comment]
```
 bloaty out/linux-x64-door-lock/chip-door-lock-app -- out-master/linux-x64-door-lock/chip-door-lock-app
    FILE SIZE        VM SIZE
 --------------  --------------
  +0.0% +4.90Ki  [ = ]       0    .debug_info
  +0.0%    +672  +0.0%    +672    .text
  +0.0%    +514  [ = ]       0    .debug_line
  +0.0%    +499  [ = ]       0    .debug_str
  +0.0%    +372  [ = ]       0    .strtab
  +0.0%    +202  [ = ]       0    .debug_abbrev
  +0.0%    +160  [ = ]       0    .debug_frame
  +0.0%     +96  [ = ]       0    .symtab
  +0.0%     +64  [ = ]       0    .debug_aranges
  +0.0%     +61  [ = ]       0    .debug_ranges
  -1.0%      -4  [ = ]       0    .shstrtab
  -9.0%    -672  [ = ]       0    [Unmapped]
  +0.0% +6.82Ki  +0.0%    +672    TOTAL
  ```
  
  Doing this because bloat CI seems to not want to update this PR yet.

**Pull request merged**